evm.md

```diff
@@ -289,48 +284,49 @@ When the `#next` operator cannot lookup the next opcode, it assumes that the end
 
 ### Single Step
 
-The `#next` operator executes a single step by:
+If the program-counter points to an actual opcode, it's loaded into the `#next [_]` operator.
+The `#next [_]` operator initiates execution by:
 
-1.  performing some quick checks for exceptional opcodes,
-2.  executes the opcode if it is not immediately exceptional,
-3.  increments the program counter, and finally
-4.  reverts state if any of the above steps threw an exception.
+1.  checking if there will be a stack over/underflow, or a static mode violation,
+2.  loading any additional state needed (when executing in full-node mode),
+3.  executing the opcode (which includes any gas deduction needed), and
+4.  adjusting the program counter.
 
 ```k
-    rule <mode> EXECMODE </mode>
-         <k> #next
-          => #stackNeeded? [ OP ]
-          ~> #static?      [ OP ]
-          ~> #load         [ OP ]
-          ~> #exec         [ OP ]
-          ~> #pc           [ OP ]
+    syntax InternalOp ::= "#next" "[" OpCode "]"
+ // --------------------------------------------
+    rule <k> #next [ OP ]
+          => #load [ OP ]
+          ~> #exec [ OP ]
+          ~> #pc   [ OP ]
          ...
          </k>
-         <pc> PCOUNT </pc>
-         <program> ... PCOUNT |-> OP ... </program>
-      requires EXECMODE in (SetItem(NORMAL) SetItem(VMTESTS))
-```
-
-### Exceptional Checks
-
--   `#stackNeeded?` checks that the stack will be not be under/overflown.
--   `#stackNeeded`, `#stackAdded`, and `#stackDelta` are helpers for deciding `#stackNeeded?`.
+         <wordStack> WS </wordStack>
+         <static> STATIC:Bool </static>
+      requires notBool ( #stackUnderflow(WS, OP) orBool #stackOverflow(WS, OP) )
+       andBool notBool ( STATIC andBool #changesState(OP, WS) )
 
-```k
-    syntax InternalOp ::= "#stackNeeded?" "[" OpCode "]"
- // ----------------------------------------------------
-    rule <k> #stackNeeded? [ OP ] => #end EVMC_STACK_UNDERFLOW ... </k>
+    rule <k> #next [ OP ] => #end EVMC_STACK_UNDERFLOW ... </k>
          <wordStack> WS </wordStack>
       requires #stackUnderflow(WS, OP)
 
-    rule <k> #stackNeeded? [ OP ] => #end EVMC_STACK_OVERFLOW ... </k>
+    rule <k> #next [ OP ] => #end EVMC_STACK_OVERFLOW ... </k>
          <wordStack> WS </wordStack>
       requires #stackOverflow(WS, OP)
 
-    rule <k> #stackNeeded? [ OP ] => . ... </k>
+    rule <k> #next [ OP ] => #end EVMC_STATIC_MODE_VIOLATION ... </k>
          <wordStack> WS </wordStack>
-      requires notBool ( #stackUnderflow(WS, OP) orBool #stackOverflow(WS, OP) )
```

Comment:
I'm fairly sure you still need this, otherwise both rules could apply if the user is doing a STATICCALL and they try to change state but have a stack exception.

Reply:
Fixed.